Increase performance

Please pass chunk of files to TSLint
or run nodejs once and call tslint in one nodejs process like gulp

error]com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Unterminated array at line 2 column 2 path $[0]
 
2016-06-23T09:17:30.0659943Z ##[error]at com.google.gson.Gson.fromJson(Gson.java:820)
 
2016-06-23T09:17:30.0659943Z ##[error]at com.google.gson.Gson.fromJson(Gson.java:770)
 
2016-06-23T09:17:30.0669709Z ##[error]at com.google.gson.Gson.fromJson(Gson.java:719)
 
2016-06-23T09:17:30.0669709Z ##[error]at com.google.gson.Gson.fromJson(Gson.java:691)
 
2016-06-23T09:17:30.0669709Z ##[error]at com.pablissimo.sonar.TsLintParserImpl.parse(TsLintParserImpl.java:12)
 
2016-06-23T09:17:30.0679475Z ##[error]at com.pablissimo.sonar.TsLintSensor.analyse(TsLintSensor.java:107)
 
2016-06-23T09:17:30.0679475Z ##[error]at org.sonar.batch.phases.SensorsExecutor.executeSensor(SensorsExecutor.java:58)
 
2016-06-23T09:17:30.0679475Z ##[error]at org.sonar.batch.phases.SensorsExecutor.execute(SensorsExecutor.java:50)
 
2016-06-23T09:17:30.0689241Z ##[error]at org.sonar.batch.phases.AbstractPhaseExecutor.execute(AbstractPhaseExecutor.java:83)
 
2016-06-23T09:17:30.0689241Z ##[error]at org.sonar.batch.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:192)
 
2016-06-23T09:17:30.0689241Z ##[error]at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
 
2016-06-23T09:17:30.0699007Z ##[error]at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
 
2016-06-23T09:17:30.0699007Z ##[error]at org.sonar.batch.scan.ProjectScanContainer.scan(ProjectScanContainer.java:241)
 
2016-06-23T09:17:30.0699007Z ##[error]at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:236)
 
2016-06-23T09:17:30.0708773Z ##[error]at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:234)
 
2016-06-23T09:17:30.0708773Z ##[error]at org.sonar.batch.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:226)
 
2016-06-23T09:17:30.0708773Z ##[error]at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
 
2016-06-23T09:17:30.0708773Z ##[error]at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
 
2016-06-23T09:17:30.0708773Z ##[error]at org.sonar.batch.task.ScanTask.execute(ScanTask.java:47)
 
2016-06-23T09:17:30.0718539Z ##[error]at org.sonar.batch.task.TaskContainer.doAfterStart(TaskContainer.java:86)
 
2016-06-23T09:17:30.0728305Z ##[error]at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
 
2016-06-23T09:17:30.0728305Z ##[error]at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
 
2016-06-23T09:17:30.0738071Z ##[error]at org.sonar.batch.bootstrap.GlobalContainer.executeTask(GlobalContainer.java:106)
 
2016-06-23T09:17:30.0738071Z ##[error]at org.sonar.batch.bootstrapper.Batch.executeTask(Batch.java:119)
 
2016-06-23T09:17:30.0738071Z ##[error]at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:62)
 
2016-06-23T09:17:30.0747837Z ##[error]at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
 
2016-06-23T09:17:30.0747837Z ##[error]at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
 
2016-06-23T09:17:30.0747837Z ##[error]at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
 
2016-06-23T09:17:30.0767369Z ##[error]at java.lang.reflect.Method.invoke(Unknown Source)
 
2016-06-23T09:17:30.0767369Z ##[error]at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
 
2016-06-23T09:17:30.0767369Z ##[error]at com.sun.proxy.$Proxy0.execute(Unknown Source)
 
2016-06-23T09:17:30.0767369Z ##[error]at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:240)
 
2016-06-23T09:17:30.0767369Z ##[error]at org.sonarsource.scanner.api.EmbeddedScanner.runAnalysis(EmbeddedScanner.java:151)
 
2016-06-23T09:17:30.0767369Z ##[error]at org.sonarsource.scanner.cli.Main.runAnalysis(Main.java:110)
 
2016-06-23T09:17:30.0777135Z ##[error]at org.sonarsource.scanner.cli.Main.execute(Main.java:72)
 
2016-06-23T09:17:30.0777135Z ##[error]at org.sonarsource.scanner.cli.Main.main(Main.java:60)
 
2016-06-23T09:17:30.0777135Z ##[error]Caused by: com.google.gson.stream.MalformedJsonException: Unterminated array at line 2 column 2 path $[0]
 
2016-06-23T09:17:30.0786901Z ##[error]at com.google.gson.stream.JsonReader.syntaxError(JsonReader.java:1572)
 
2016-06-23T09:17:30.0786901Z ##[error]at com.google.gson.stream.JsonReader.doPeek(JsonReader.java:478)
 
2016-06-23T09:17:30.0786901Z ##[error]at com.google.gson.stream.JsonReader.hasNext(JsonReader.java:416)
 
2016-06-23T09:17:30.0796667Z ##[error]at com.google.gson.internal.bind.ArrayTypeAdapter.read(ArrayTypeAdapter.java:71)
 
2016-06-23T09:17:30.0796667Z ##[error]at com.google.gson.Gson.fromJson(Gson.java:805)

--- a/.gitignore
+++ b/.gitignore
@@ -48,3 +48,4 @@ hs_err_pid*
 
 #Custom stuff
 sample.tslint.json
+.idea

--- a/pom.xml
+++ b/pom.xml
@@ -5,7 +5,7 @@
   <groupId>com.pablissimo.sonar</groupId>
   <artifactId>sonar-typescript-plugin</artifactId>
   <packaging>sonar-plugin</packaging>
-  <version>0.4-SNAPSHOT</version>
+  <version>0.5-SNAPSHOT</version>
 
   <name>TypeScript</name>
   <description>Analyse TypeScript projects</description>
@@ -14,7 +14,7 @@
     <connection>scm:git:git@github.com:Pablissimo/SonarTsPlugin.git</connection>
     <developerConnection>scm:git:git@github.com:Pablissimo/SonarTsPlugin.git</developerConnection>
     <url>https://github.com/Pablissimo/SonarTsPlugin</url>
-    <tag>0.4</tag>
+    <tag>0.5</tag>
   </scm>
   
   <properties>

--- a/pom.xml
+++ b/pom.xml
@@ -5,7 +5,7 @@
   <groupId>com.pablissimo.sonar</groupId>
   <artifactId>sonar-typescript-plugin</artifactId>
   <packaging>sonar-plugin</packaging>
-  <version>0.5-SNAPSHOT</version>
+  <version>0.6-SNAPSHOT</version>
 
   <name>TypeScript</name>
   <description>Analyse TypeScript projects</description>
@@ -14,7 +14,7 @@
     <connection>scm:git:git@github.com:Pablissimo/SonarTsPlugin.git</connection>
     <developerConnection>scm:git:git@github.com:Pablissimo/SonarTsPlugin.git</developerConnection>
     <url>https://github.com/Pablissimo/SonarTsPlugin</url>
-    <tag>0.5</tag>
+    <tag>0.6</tag>
   </scm>
   
   <properties>

--- a/src/main/java/com/pablissimo/sonar/TsLintExecutor.java
+++ b/src/main/java/com/pablissimo/sonar/TsLintExecutor.java
@@ -1,5 +1,7 @@
 package com.pablissimo.sonar;
 
+import java.util.List;
+
 public interface TsLintExecutor {
-    String execute(String pathToTsLint, String configFile, String rulesDir, String file, Integer timeoutMs);
+    String execute(String pathToTsLint, String configFile, String rulesDir, List<String> files, Integer timeoutMs);
 }

--- a/src/main/java/com/pablissimo/sonar/TsLintExecutorImpl.java
+++ b/src/main/java/com/pablissimo/sonar/TsLintExecutorImpl.java
@@ -1,5 +1,8 @@
 package com.pablissimo.sonar;
 
+import java.util.ArrayList;
+import java.util.List;
+
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.sonar.api.utils.command.Command;
@@ -7,28 +10,62 @@ import org.sonar.api.utils.command.CommandExecutor;
 import org.sonar.api.utils.command.StreamConsumer;
 
 public class TsLintExecutorImpl implements TsLintExecutor {
+    public static final int MAX_COMMAND_LENGTH = 4096;
     private static final Logger LOG = LoggerFactory.getLogger(TsLintExecutorImpl.class);
-
+    
     private StringBuilder stdOut;
     private StringBuilder stdErr;
 
-    public String execute(String pathToTsLint, String configFile, String rulesDir, String file, Integer timeoutMs) {
-        LOG.info("TsLint executing for " + file);
-        Command command = Command.create("node");
-        command.addArgument(pathToTsLint);
-        command.addArgument("--format");
-        command.addArgument("json");
+    private static Command getBaseCommand(String pathToTsLint, String configFile, String rulesDir) {
+        Command command = 
+                Command
+                .create("node")
+                .addArgument(pathToTsLint)
+                .addArgument("--format")
+                .addArgument("json");
         
         if (rulesDir != null && rulesDir.length() > 0) {
-            command.addArgument("--rules-dir");
-            command.addArgument(rulesDir);
+            command
+                .addArgument("--rules-dir")
+                .addArgument(rulesDir);
         }
         
-        command.addArgument("--config");
-        command.addArgument(configFile);
-        command.addArgument(file.trim());
-        command.setNewShell(false);
-
+        command
+            .addArgument("--config")
+            .addArgument(configFile)
+            .setNewShell(false);
+            
+        return command;
+    }
+    
+    public String execute(String pathToTsLint, String configFile, String rulesDir, List<String> files, Integer timeoutMs) {
+        // New up a command that's everything we need except the files to process
+        // We'll use this as our reference for chunking up files
+        int baseCommandLength = getBaseCommand(pathToTsLint, configFile, rulesDir).toCommandLine().length();
+        int availableForBatching = MAX_COMMAND_LENGTH - baseCommandLength;
+        
+        List<List<String>> batches = new ArrayList<List<String>>();
+        List<String> currentBatch = new ArrayList<String>();
+        batches.add(currentBatch);
+        
+        int currentBatchLength = 0;
+        for (int i = 0; i < files.size(); i++) {
+            String nextPath = files.get(i).trim();
+            
+            // +1 for the space we'll be adding between filenames
+            if (currentBatchLength + nextPath.length() + 1 > availableForBatching) { 
+                // Too long to add to this batch, create new
+                currentBatch = new ArrayList<String>();
+                currentBatchLength = 0;
+                batches.add(currentBatch);
+            }
+            
+            currentBatch.add(nextPath);
+            currentBatchLength += nextPath.length() + 1;
+        }
+        
+        LOG.debug("Split " + files.size() + " files into " + batches.size() + " batches for processing");
+        
         this.stdOut = new StringBuilder();
         this.stdErr = new StringBuilder();
 
@@ -46,9 +83,26 @@ public class TsLintExecutorImpl implements TsLintExecutor {
             }
         };
 
-        this.createExecutor().execute(command, stdOutConsumer, stdErrConsumer, timeoutMs);
-
-        return stdOut.toString();
+        for (int i = 0; i < batches.size(); i++) {
+            List<String> thisBatch = batches.get(i);
+            Command thisCommand = getBaseCommand(pathToTsLint, configFile, rulesDir);
+            
+            for (int fileIndex = 0; fileIndex < thisBatch.size(); fileIndex++) {
+                thisCommand.addArgument(thisBatch.get(fileIndex));
+            }
+            
+            LOG.debug("Executing TsLint with command: " + thisCommand.toCommandLine());
+            
+            // Timeout is specified per file, not per batch (which can vary a lot)
+            // so multiply it up
+            this.createExecutor().execute(thisCommand, stdOutConsumer, stdErrConsumer, timeoutMs * thisBatch.size());
+        }
+        
+        String rawOutput = stdOut.toString();
+        
+        // TsLint returns nonsense for its JSON output when faced with multiple files
+        // so we need to fix it up before we do anything else
+        return "[" + rawOutput.replaceAll("\\]\\[", "],[") + "]";
     }
     
     protected CommandExecutor createExecutor() {

--- a/src/main/java/com/pablissimo/sonar/TsLintExecutorImpl.java
+++ b/src/main/java/com/pablissimo/sonar/TsLintExecutorImpl.java
@@ -12,67 +12,67 @@ import org.sonar.api.utils.command.StreamConsumer;
 public class TsLintExecutorImpl implements TsLintExecutor {
     public static final int MAX_COMMAND_LENGTH = 4096;
     private static final Logger LOG = LoggerFactory.getLogger(TsLintExecutorImpl.class);
-    
+
     private StringBuilder stdOut;
     private StringBuilder stdErr;
 
     private static Command getBaseCommand(String pathToTsLint, String configFile, String rulesDir) {
-        Command command = 
+        Command command =
                 Command
                 .create("node")
                 .addArgument(pathToTsLint)
                 .addArgument("--format")
                 .addArgument("json");
-        
+
         if (rulesDir != null && rulesDir.length() > 0) {
             command
                 .addArgument("--rules-dir")
                 .addArgument(rulesDir);
         }
-        
+
         command
             .addArgument("--config")
             .addArgument(configFile)
             .setNewShell(false);
-            
+
         return command;
     }
-    
+
     public String execute(String pathToTsLint, String configFile, String rulesDir, List<String> files, Integer timeoutMs) {
         // New up a command that's everything we need except the files to process
         // We'll use this as our reference for chunking up files
         int baseCommandLength = getBaseCommand(pathToTsLint, configFile, rulesDir).toCommandLine().length();
         int availableForBatching = MAX_COMMAND_LENGTH - baseCommandLength;
-        
+
         List<List<String>> batches = new ArrayList<List<String>>();
         List<String> currentBatch = new ArrayList<String>();
         batches.add(currentBatch);
-        
+
         int currentBatchLength = 0;
         for (int i = 0; i < files.size(); i++) {
             String nextPath = files.get(i).trim();
-            
+
             // +1 for the space we'll be adding between filenames
-            if (currentBatchLength + nextPath.length() + 1 > availableForBatching) { 
+            if (currentBatchLength + nextPath.length() + 1 > availableForBatching) {
                 // Too long to add to this batch, create new
                 currentBatch = new ArrayList<String>();
                 currentBatchLength = 0;
                 batches.add(currentBatch);
             }
-            
+
             currentBatch.add(nextPath);
             currentBatchLength += nextPath.length() + 1;
         }
-        
+
         LOG.debug("Split " + files.size() + " files into " + batches.size() + " batches for processing");
-        
+
         this.stdOut = new StringBuilder();
         this.stdErr = new StringBuilder();
 
         StreamConsumer stdOutConsumer = new StreamConsumer() {
             public void consumeLine(String line) {
                 LOG.trace("TsLint Out: " + line);
-                stdOut.append(line + "\n");
+                stdOut.append(line);
             }
         };
 
@@ -86,25 +86,25 @@ public class TsLintExecutorImpl implements TsLintExecutor {
         for (int i = 0; i < batches.size(); i++) {
             List<String> thisBatch = batches.get(i);
             Command thisCommand = getBaseCommand(pathToTsLint, configFile, rulesDir);
-            
+
             for (int fileIndex = 0; fileIndex < thisBatch.size(); fileIndex++) {
                 thisCommand.addArgument(thisBatch.get(fileIndex));
             }
-            
+
             LOG.debug("Executing TsLint with command: " + thisCommand.toCommandLine());
-            
+
             // Timeout is specified per file, not per batch (which can vary a lot)
             // so multiply it up
             this.createExecutor().execute(thisCommand, stdOutConsumer, stdErrConsumer, timeoutMs * thisBatch.size());
         }
-        
+
         String rawOutput = stdOut.toString();
-        
+
         // TsLint returns nonsense for its JSON output when faced with multiple files
         // so we need to fix it up before we do anything else
         return "[" + rawOutput.replaceAll("\\]\\[", "],[") + "]";
     }
-    
+
     protected CommandExecutor createExecutor() {
         return CommandExecutor.create();
     }

--- a/src/main/java/com/pablissimo/sonar/TsLintParser.java
+++ b/src/main/java/com/pablissimo/sonar/TsLintParser.java
@@ -3,5 +3,5 @@ package com.pablissimo.sonar;
 import com.pablissimo.sonar.model.TsLintIssue;
 
 public interface TsLintParser {
-    TsLintIssue[] parse(String toParse);
+    TsLintIssue[][] parse(String toParse);
 }

--- a/src/main/java/com/pablissimo/sonar/TsLintParserImpl.java
+++ b/src/main/java/com/pablissimo/sonar/TsLintParserImpl.java
@@ -5,10 +5,10 @@ import com.google.gson.GsonBuilder;
 import com.pablissimo.sonar.model.TsLintIssue;
 
 public class TsLintParserImpl implements TsLintParser {
-    public TsLintIssue[] parse(String toParse) {
+    public TsLintIssue[][] parse(String toParse) {
         GsonBuilder builder = new GsonBuilder();
         Gson gson = builder.create();
 
-        return gson.fromJson(toParse, TsLintIssue[].class);
+        return gson.fromJson(toParse, TsLintIssue[][].class);
     }
 }

--- a/src/main/java/com/pablissimo/sonar/TsLintSensor.java
+++ b/src/main/java/com/pablissimo/sonar/TsLintSensor.java
@@ -5,6 +5,7 @@ import java.io.IOException;
 import java.nio.charset.Charset;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 
@@ -90,37 +91,62 @@ public class TsLintSensor implements Sensor {
             ruleNames.add(rule.getKey());
         }
 
+        List<String> paths = new ArrayList<String>();
+        HashMap<String, File> fileMap = new HashMap<String, File>();
+        
         for (File file : fileSystem.files(this.filePredicates.hasLanguage(TypeScriptLanguage.LANGUAGE_KEY))) {
             if (skipTypeDefFiles && file.getName().toLowerCase().endsWith("." + TypeScriptLanguage.LANGUAGE_DEFINITION_EXTENSION)) {
                 continue;
             }
-
+            
+            String pathAdjusted = file.getAbsolutePath().replace('\\', '/');
+            paths.add(pathAdjusted);
+            fileMap.put(pathAdjusted, file);
+        }
+        
+        String jsonResult = executor.execute(pathToTsLint, pathToTsLintConfig, rulesDir, paths, tsLintTimeoutMs);
+                
+        TsLintIssue[][] issues = parser.parse(jsonResult);
+        
+        if (issues == null) {
+            LOG.warn("TsLint returned no result at all");
+            return;
+        }
+        
+        // Each issue bucket will contain info about a single file
+        for (TsLintIssue[] batchIssues : issues) {
+            if (batchIssues == null || batchIssues.length == 0) {
+                continue;
+            }
+            
+            String filePath = batchIssues[0].getName();
+            
+            if (!fileMap.containsKey(filePath)) {
+                LOG.warn("TsLint reported issues against a file that wasn't sent to it - will be ignored: " + filePath);
+                continue;
+            }
+            
+            File file = fileMap.get(filePath);
             Resource resource = this.getFileFromIOFile(file, project);
             Issuable issuable = perspectives.as(Issuable.class, resource);
-
-            String jsonResult = executor.execute(pathToTsLint, pathToTsLintConfig, rulesDir, file.getAbsolutePath(), tsLintTimeoutMs);
-
-            TsLintIssue[] issues = parser.parse(jsonResult);
-
-            if (issues != null) {
-                for (TsLintIssue issue : issues) {
-                    // Make sure the rule we're violating is one we recognise - if not, we'll
-                    // fall back to the generic 'tslint-issue' rule
-                    String ruleName = issue.getRuleName();
-                    if (!ruleNames.contains(ruleName)) {
-                        ruleName = TsRulesDefinition.RULE_TSLINT_ISSUE;
-                    }
-
-                    issuable.addIssue
-                    (
-                            issuable
-                            .newIssueBuilder()
-                            .line(issue.getStartPosition().getLine() + 1)
-                            .message(issue.getFailure())
-                            .ruleKey(RuleKey.of(TsRulesDefinition.REPOSITORY_NAME, ruleName))
-                            .build()
-                            );
+            
+            for (TsLintIssue issue : batchIssues) {
+                // Make sure the rule we're violating is one we recognise - if not, we'll
+                // fall back to the generic 'tslint-issue' rule
+                String ruleName = issue.getRuleName();
+                if (!ruleNames.contains(ruleName)) {
+                    ruleName = TsRulesDefinition.RULE_TSLINT_ISSUE;
                 }
+    
+                issuable.addIssue
+                (
+                        issuable
+                        .newIssueBuilder()
+                        .line(issue.getStartPosition().getLine() + 1)
+                        .message(issue.getFailure())
+                        .ruleKey(RuleKey.of(TsRulesDefinition.REPOSITORY_NAME, ruleName))
+                        .build()
+                        );
             }
         }
     }

--- a/src/test/java/com/pablissimo/sonar/TsLintExecutorImplTest.java
+++ b/src/test/java/com/pablissimo/sonar/TsLintExecutorImplTest.java
@@ -4,6 +4,8 @@ import static org.junit.Assert.*;
 import static org.mockito.Mockito.*;
 
 import java.util.ArrayList;
+import java.util.List;
+
 import org.junit.Before;
 import org.junit.Test;
 import org.mockito.invocation.InvocationOnMock;
@@ -11,6 +13,8 @@ import org.mockito.stubbing.Answer;
 import org.sonar.api.utils.command.Command;
 import org.sonar.api.utils.command.CommandExecutor;
 import org.sonar.api.utils.command.StreamConsumer;
+
+import edu.emory.mathcs.backport.java.util.Arrays;
 
 public class TsLintExecutorImplTest {
     TsLintExecutorImpl executorImpl;
@@ -24,7 +28,7 @@ public class TsLintExecutorImplTest {
     }
     
     @Test
-    public void executesCommandWithCorrectArguments() {
+    public void executesCommandWithCorrectArgumentsAndTimeouts() {
         final ArrayList<Command> capturedCommands = new ArrayList<Command>();
         final ArrayList<Long> capturedTimeouts = new ArrayList<Long>();
         
@@ -38,14 +42,93 @@ public class TsLintExecutorImplTest {
         };
         
         when(this.commandExecutor.execute(any(Command.class), any(StreamConsumer.class), any(StreamConsumer.class), any(long.class))).then(captureCommand);
-        this.executorImpl.execute("path/to/tslint", "path/to/config", "path/to/rules", "path/to/file", 40000);
+        this.executorImpl.execute("path/to/tslint", "path/to/config", "path/to/rules", Arrays.asList(new String[] { "path/to/file", "path/to/another" }), 40000);
         
         assertEquals(1, capturedCommands.size());
         
         Command theCommand = capturedCommands.get(0);
         long theTimeout = capturedTimeouts.get(0);
         
-        assertEquals("node path/to/tslint --format json --rules-dir path/to/rules --config path/to/config path/to/file", theCommand.toCommandLine());
-        assertEquals(40000, theTimeout);        
+        assertEquals("node path/to/tslint --format json --rules-dir path/to/rules --config path/to/config path/to/file path/to/another", theCommand.toCommandLine());
+        // Expect one timeout period per file processed
+        assertEquals(2 * 40000, theTimeout);        
+    }
+    
+    @Test
+    public void DoesNotAddRulesDirParameter_IfNull() {
+        final ArrayList<Command> capturedCommands = new ArrayList<Command>();
+        final ArrayList<Long> capturedTimeouts = new ArrayList<Long>();
+        
+        Answer<Integer> captureCommand = new Answer<Integer>() {
+            @Override
+            public Integer answer(InvocationOnMock invocation) throws Throwable {
+                capturedCommands.add((Command) invocation.getArguments()[0]);
+                capturedTimeouts.add((long) invocation.getArguments()[3]);
+                return 0;
+            }
+        };
+        
+        when(this.commandExecutor.execute(any(Command.class), any(StreamConsumer.class), any(StreamConsumer.class), any(long.class))).then(captureCommand);
+        this.executorImpl.execute("path/to/tslint", "path/to/config", null, Arrays.asList(new String[] { "path/to/file" }), 40000);
+        
+        Command theCommand = capturedCommands.get(0);
+        assertFalse(theCommand.toCommandLine().contains("--rules-dir"));
+    }
+    
+    @Test
+    public void DoesNotAddRulesDirParameter_IfEmptyString() {
+        final ArrayList<Command> capturedCommands = new ArrayList<Command>();
+        final ArrayList<Long> capturedTimeouts = new ArrayList<Long>();
+        
+        Answer<Integer> captureCommand = new Answer<Integer>() {
+            @Override
+            public Integer answer(InvocationOnMock invocation) throws Throwable {
+                capturedCommands.add((Command) invocation.getArguments()[0]);
+                capturedTimeouts.add((long) invocation.getArguments()[3]);
+                return 0;
+            }
+        };
+        
+        when(this.commandExecutor.execute(any(Command.class), any(StreamConsumer.class), any(StreamConsumer.class), any(long.class))).then(captureCommand);
+        this.executorImpl.execute("path/to/tslint", "path/to/config", "", Arrays.asList(new String[] { "path/to/file" }), 40000);
+        
+        Command theCommand = capturedCommands.get(0);
+        assertFalse(theCommand.toCommandLine().contains("--rules-dir"));
+    }
+    
+    @Test
+    public void BatchesExecutions_IfTooManyFilesForCommandLine() {
+        List<String> filenames = new ArrayList<String>();
+        int currentLength = 0;
+        int standardCmdLength = "node path/to/tslint --format json --rules-dir path/to/rules --config path/to/config".length();
+        
+        String firstBatch = "first batch";
+        while (currentLength + 12 < TsLintExecutorImpl.MAX_COMMAND_LENGTH - standardCmdLength) {
+            filenames.add(firstBatch);
+            currentLength += firstBatch.length() + 1;
+        }
+        filenames.add("second batch");
+        
+        final ArrayList<Command> capturedCommands = new ArrayList<Command>();
+        final ArrayList<Long> capturedTimeouts = new ArrayList<Long>();
+        
+        Answer<Integer> captureCommand = new Answer<Integer>() {
+            @Override
+            public Integer answer(InvocationOnMock invocation) throws Throwable {
+                capturedCommands.add((Command) invocation.getArguments()[0]);
+                capturedTimeouts.add((long) invocation.getArguments()[3]);
+                return 0;
+            }
+        };
+        
+        when(this.commandExecutor.execute(any(Command.class), any(StreamConsumer.class), any(StreamConsumer.class), any(long.class))).then(captureCommand);
+        this.executorImpl.execute("path/to/tslint", "path/to/config", "path/to/rules", filenames, 40000);
+        
+        assertEquals(2, capturedCommands.size());
+        
+        Command theSecondCommand = capturedCommands.get(1);
+        
+        assertFalse(theSecondCommand.toCommandLine().contains("first batch"));
+        assertTrue(theSecondCommand.toCommandLine().contains("second batch"));
     }
 }

--- a/src/test/java/com/pablissimo/sonar/TsLintParserTest.java
+++ b/src/test/java/com/pablissimo/sonar/TsLintParserTest.java
@@ -16,12 +16,16 @@ public class TsLintParserTest {
 
     @Test
     public void parsesValidTsLintRecordIntoObject() {
-        String toParse = "[{\"endPosition\":{\"character\":44,\"line\":23,\"position\":658},\"failure\":\"for statements must be braced\",\"name\":\"Tools.ts\",\"ruleName\":\"curly\",\"startPosition\":{\"character\":6,\"line\":22,\"position\":587}}]";
-        TsLintIssue[] issues = new TsLintParserImpl().parse(toParse);
+        String toParse = "[[{\"endPosition\":{\"character\":44,\"line\":23,\"position\":658},\"failure\":\"for statements must be braced\",\"name\":\"Tools.ts\",\"ruleName\":\"curly\",\"startPosition\":{\"character\":6,\"line\":22,\"position\":587}}]]";
+        TsLintIssue[][] issues = new TsLintParserImpl().parse(toParse);
 
         assertEquals(1,  issues.length);
 
-        TsLintIssue issue = issues[0];
+        TsLintIssue[] fileIssues = issues[0];
+        
+        assertEquals(1, fileIssues.length);
+        
+        TsLintIssue issue = fileIssues[0];
         assertEquals(44, issue.getEndPosition().getCharacter());
         assertEquals(23, issue.getEndPosition().getLine());
         assertEquals(658, issue.getEndPosition().getPosition());

--- a/src/test/java/com/pablissimo/sonar/TsLintSensorTest.java
+++ b/src/test/java/com/pablissimo/sonar/TsLintSensorTest.java
@@ -63,6 +63,7 @@ public class TsLintSensorTest {
 
         this.file = mock(File.class);
         doReturn(true).when(this.file).isFile();
+        doReturn("/path/to/file").when(this.file).getAbsolutePath();
 
         this.files = new ArrayList<File>(Arrays.asList(new File[] {
                 this.file
@@ -106,14 +107,15 @@ public class TsLintSensorTest {
         TsLintIssue issue = new TsLintIssue();
         issue.setFailure("failure");
         issue.setRuleName("rule name");
+        issue.setName("/path/to/file");
 
         TsLintPosition startPosition = new TsLintPosition();
         startPosition.setLine(1);
 
         issue.setStartPosition(startPosition);
 
-        TsLintIssue[] issues = new TsLintIssue[] {
-                issue
+        TsLintIssue[][] issues = new TsLintIssue[][] {
+                new TsLintIssue[] { issue }
         };
 
         final List<Issue> capturedIssues = new ArrayList<Issue>();
@@ -155,7 +157,7 @@ public class TsLintSensorTest {
     public void analyse_callsExecutorWithSuppliedTimeout() throws IOException {
         this.sensor.analyse(mock(Project.class), mock(SensorContext.class));
      
-        verify(this.executor, times(1)).execute(any(String.class), any(String.class), any(String.class), any(String.class), eq(45000));
+        verify(this.executor, times(1)).execute(any(String.class), any(String.class), any(String.class), any(List.class), eq(45000));
     }
     
     @Test
@@ -164,6 +166,6 @@ public class TsLintSensorTest {
         
         this.sensor.analyse(mock(Project.class), mock(SensorContext.class));
         
-        verify(this.executor, times(1)).execute(any(String.class), any(String.class), any(String.class), any(String.class), eq(5000));
+        verify(this.executor, times(1)).execute(any(String.class), any(String.class), any(String.class), any(List.class), eq(5000));
     }
 }

